feat: Adds the ECharts Bubble chart

mayurnewase opened the pull request:
### SUMMARY
Adds the ECharts Bubble chart.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
1. Chart and data control panel:
![Screenshot from 2022-11-23 10-52-20](https://user-images.githubusercontent.com/12967587/203477116-08c4fa72-daac-4465-9290-2511eac70139.png)

2. Customize control panel
![Screenshot from 2022-11-27 16-18-56](https://user-images.githubusercontent.com/12967587/204131304-8b142e5e-acd2-436f-acb9-7d3f2aa66f94.png)

3. Tooltip:
![Screenshot from 2022-11-23 11-11-23](https://user-images.githubusercontent.com/12967587/203477703-60049948-5e74-47e7-a06f-b7180f323a8b.png)

### TESTING INSTRUCTIONS
1. create d3 bubble chart
2. merge commits from this PR and open existing chart should work fine as before.
4. test all controls in `data` and `customize` sections.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

mayurnewase committed:
initial

mayurnewase committed:
fix types and add story

mayurnewase committed:
add tests for buildquery and transformprops

mayurnewase committed:
remove console

mayurnewase committed:
replace chart plugin, add license in new files

mayurnewase committed:
more lint more license more fun

mayurnewase commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,183 @@
+    ? `${params.data[3]} <sub>(${params.data[4]})</sub>`
+    : params.data[3];
+
+  return `<p>${title}</p>
```
didn't use echart's sanitize function, it removes line breaks.

mayurnewase commented on superset-frontend/.storybook/main.js:
```diff
@@ -24,6 +24,7 @@ module.exports = {
     builder: 'webpack5',
   },
   stories: [
+    '../packages/superset-ui-demo/storybook/**/Stories.@(tsx|jsx|mdx)',
```
bycatch: should include stories for plugins when storybook server is run.

michael-s-molina replied:
Actually they are separate storybooks currently. Before merging them, we need to think about how to organize the components.

mayurnewase commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/index.ts:
```diff
@@ -0,0 +1,57 @@
+        description: t(
+          'Visualizes a metric across three dimensions of data in a single chart (X axis, Y axis, and bubble size). Bubbles from the same group can be showcased using bubble color.',
+        ),
+        name: t('Bubble Chart V2'),
```
will revert to original `Bubble Chart` in next commit.

mayurnewase commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,183 @@
+    // eslint-disable-next-line no-param-reassign
+    node.symbolSize =
+      (((node.data![0][2] - bubbleMinValue) / nodeSpread) *
+        (maxBubbleValue * 2) || 0) + MINIMUM_BUBBLE_SIZE;
```
`* 2` to have same size options as previous charts in the control panel to avoid db migration.

mayurnewase committed:
rename

mayurnewase committed:
update

mayurnewase committed:
lint

mayurnewase committed:
fix import

villebro commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,188 @@
+  const legends: string[] = [];
+  const series: ScatterSeriesOption[] = [];
+
+  const xAxisLabel: string = x.label || x;
+  const yAxisLabel: string = y.label || y;
+  const sizeLabel: string = size.label || size;
```
This will only work for adhoc metrics, use `getMetricLabel` from `@superset-ui/core` to make it also support saved metrics: 
```suggestion
  const xAxisLabel: string = getMetricLabel(x);
  const yAxisLabel: string = getMetricLabel(y);
  const sizeLabel: string = getMetricLabel(size);
```

villebro commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,188 @@
+        [
+          datum[xAxisLabel],
+          datum[yAxisLabel],
+          datum[size.label],
```
missed this one:
```suggestion
          datum[sizeLabel],
```

villebro commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,188 @@
+    },
+    tooltip: {
+      show: !inContextMenu,
+      ...defaultTooltip,
```
`defaultTooltip` seems to have been a leftover from a recent fix (my bad; I should have removed it in #22218). Please use `getDefaultTooltip` from `utils/tooltip.ts` (this ensures proper tooltip positioning) and remove `defaultTooltip` in `defaults.ts`:
```suggestion
      ...getDefaultTooltip(refs),
```

villebro commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,188 @@
+    : params.data[3];
+
+  return `<p>${title}</p>
+        ${xAxisLabel}: ${params.data[0]} <br/>
+        ${yAxisLabel}: ${params.data[1]} <br/>
+        ${sizeLabel}: ${params.data[2]}`;
```
We should probably use the formatters here for x and y values. In some other plugins I believe we have separate formatters for the axes and the tooltip - I think that could be appropriate here. Also, should we perhaps have a formatter for the bubble size, too?

mayurnewase replied:
should we have single formatter for all 3 values?

mayurnewase replied:
having 3 seaparate seems bit overkill.

mayurnewase replied:
added 1, let me know if we need 3 separate controls.

villebro replied:
Looking at the example chart in the World Bank Dashboard, the size is in billions, while t x-axis and y-axis are essentially percentages:
<img width="765" alt="image" src="https://user-images.githubusercontent.com/33317356/204468139-f738e78a-c550-4129-9efb-6c4b9b0210cb.png">

I think x and y axes will usually be in the same scale, but since we already have separate formatters for them and no separate tooltip formatter, I would maybe just propose adding a size formatter, and then using the same formatters in the tooltip for now (later on if people really want to use separate formatters on the tooltip we can add that, but let's not go for 6 formatters quite yet 😬).

mayurnewase replied:
added.

villebro reviewed (commented):
First pass comments, looks great! One thing that I noticed is that the NVD3 plugin assumes truncation on the y-axis if the bounds aren't set. While I don't necessarily like the idea of defaulting to truncation (actually, I strongly oppose it), I was just wondering if we should do a migration that enables truncation for existing charts if the bounds are undefined?

mayurnewase committed:
fix bubble cypress integration test

mayurnewase committed:
add opacity and tooltip formatter control, update review comments

mayurnewase committed:
reduce default opacity to 0.6

mayurnewase committed:
add migration to set y axis truncation

mayurnewase committed:
add size formatter

mayurnewase commented:
> First pass comments, looks great! One thing that I noticed is that the NVD3 plugin assumes truncation on the y-axis if the bounds aren't set. While I don't necessarily like the idea of defaulting to truncation (actually, I strongly oppose it), I was just wondering if we should do a migration that enables truncation for existing charts if the bounds are undefined?

agree, added migration.

mayurnewase committed:
fix migration conflict

villebro commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,207 @@
+  const { onContextMenu, setDataMask = () => {} } = hooks;
+
+  return {
+    height,
```
we need to pass refs here to the component:
```suggestion
    refs,
    height,
```

villebro commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/transformProps.ts:
```diff
@@ -0,0 +1,207 @@
+      max,
+      type: logYAxis ? AxisType.log : AxisType.value,
+    },
+    legend: {
+      ...getLegendProps(LegendType.Scroll, LegendOrientation.Top, true),
+      data: legends,
+    },
```
Let's make the legend customizable like in the other charts, both so it can be disabled/enabled and also changeing the orientation etc (there should be controls and utils that can be readily reused)

villebro reviewed (commented):
A few more things that came to mind. Also, one major issue: the chart is heavily padded:
<img width="923" alt="image" src="https://user-images.githubusercontent.com/33317356/204568913-e2294e89-69f0-468e-9d86-16ca740843c4.png">
I quickly verified that by adding `left: 0, right: 0` to the grid seemed to widen it to the full width, although it went a few pixels over. So I think we need to reuse the `getPadding` util from the Timeseries chart to get the grid right.

mayurnewase committed:
fix padding, add refs, improve bubble label test with different formatters

mayurnewase committed:
remove redundant description from controls

zhaoyongjie reviewed (commented):
I just saw the original bubble chart(NVD3 version), and the `limit control` have an inconsistent setting. IMO, we should fix/improve that in the __new version__ of Bubble Chart.

1. The `Series Limit` refers to a limit by series rather than a `row limit`.
2. No chance to set a `order by` clause in the SQL

![image](https://user-images.githubusercontent.com/2016594/204731203-9f7a99a5-c9a8-4f89-8844-f711dfa9bbc1.png)

zhaoyongjie commented on superset-frontend/plugins/plugin-chart-echarts/src/Bubble/constants.ts:
```diff
@@ -0,0 +1,36 @@
+import { DEFAULT_LEGEND_FORM_DATA } from '../constants';
+import { EchartsBubbleFormData } from './types';
+
+// @ts-ignore
```
Could remove the `ts-ignore` in the new plugins?

mayurnewase replied:
its because `SqlaFormData` has many values which are not set default like viz_type, datasource etc.

mayurnewase replied:
should make them optional?

mayurnewase replied:
it will need modification in core, which again needs 100% coverage which will increase the PR size.

michael-s-molina replied:
Resolved with `Partial`.

mayurnewase commented:
> I just saw the original bubble chart(NVD3 version), and the `limit control` have an inconsistent setting. IMO, we should fix/improve that in the **new version** of Bubble Chart.
> 
> 1. The `Series Limit` refers to a limit by series rather than a `row limit`.
> 2. No chance to set a `order by` clause in the SQL
> 
> ![image](https://user-images.githubusercontent.com/2016594/204731203-9f7a99a5-c9a8-4f89-8844-f711dfa9bbc1.png)

here series limit made sense as its inline with echart's options.(every bubble is separate series)
and description of `dimension` in control panel had series in it to group them by color.

mayurnewase committed:
update buildquery for columns, show dimension in tooltip on next line

mayurnewase commented:
> @mayurnewase the `series limit` will send a subquery instead a simple query, the time-series chart should be referenced.

I see, let me check.

villebro commented:
FYI @mayurnewase @zhaoyongjie , slightly related, I'm opening a PR today that will clean up deprecated limit control logic from superset-ui/core so we can remove those annoying deprecation warnings on chart data requests

mayurnewase committed:
add orderby support, fix log on axes

mayurnewase commented:
updated series limit to row limit, and added order control.

@zhaoyongjie the `normalizeOrderBy` in core doesn't support this control panel yet and I didn't want to change anything in core in this PR so added order support in the buildQuery itself.
Either this or need to something more hacky like below

`orderby: normalizeOrderBy(
      {
        ...baseQueryObject,
        orderby: [
          [...baseQueryObject.orderby, !baseQueryObject.order_desc]
        ]
      }
    ).orderby`

what do you think?

rusackas commented:
Just seeing this and catching up on the thread - this is great, thanks!

I haven't looked at the code yet, but for myself or any other reviewer, I'm hopeful that this (or a followup) includes the latest best practices:
• Generic X Axis
• Standardized controls (normalizing/denormalizing for fast-viz switching)
• Drill to Detail support

Basically, trying to keep all the ECharts plugins current with all the first-class bells and whistles. Happy to provide support/references for any of the above if needed.

Again, thanks, this is exciting!

mayurnewase commented:
> Just seeing this and catching up on the thread - this is great, thanks!
> 
> I haven't looked at the code yet, but for myself or any other reviewer, I'm hopeful that this (or a followup) includes the latest best practices: • Generic X Axis • Standardized controls (normalizing/denormalizing for fast-viz switching) • Drill to Detail support
> 
> Basically, trying to keep all the ECharts plugins current with all the first-class bells and whistles. Happy to provide support/references for any of the above if needed.
> 
> Again, thanks, this is exciting!

@rusackas 
1. x and y axes are both metrics
2. controls are standard but for fast switching don't think there is any other chart with x and y and 1 more metric for bubble size, so not sure how I can test the switch.
3. I will add cross filter and drill to detail as soon as this PR gets merged, because it may need other modifications and new tests.

possible to get 1 more review?

rusackas commented:
/testenv up

rusackas commented:
Is there any intent to support temporal or categorical X-axis on this? 

Also, the ephemeral environment is up and running if any of the earlier reviewers wouldn't mind giving this another pass.

rusackas commented:
There are a few things from the old chart that aren't supported in this new one
• Display of X bounds and Wy bounds (the extent numbers on the axes)
• Auto-rotation of axis labels
• Axis limit inputs (love 'em or hate 'em)
• Series limit (this one seems like it would be important)
• Max Bubble Size

While I think _most_ people would be fine with _most_ of the above, it might be a bit contentious to call it a drop-in replacement without complete feature parity. We've learned this lesson before and been stuck with difficult reverts. I see a few options:

1) Go for complete feature parity (which we might not want to!)
2) Post a DISCUSS thread asking for lazy consensus on the dev list, to see if others are OK with this change, and proceed accordingly
3) Add this as a new chart, mark the current one as legacy, and live with both in the product until we can make a breaking change (3.0)

I'm thinking  option 2 might be prudent. I LOVE this PR and would love to see any NVD3 chart die, but I want to make sure we're not pushing any (admittedly small) breaking changes on people without at least some form of official consent.

mayurnewase commented:
1. Display of X bounds and Wy bounds (the extent numbers on the axes) -> here bounds are visible by default.

2. Axis limit inputs -> just like old chart this one has Y axis limit -> need to click on `Truncate Y Axis`
![Screenshot from 2022-12-14 07-45-01](https://user-images.githubusercontent.com/12967587/207488720-1db6e7ab-b185-42ed-8ebe-86d6a4fc0af2.png)


3. Auto rotation -> I didnt add this for consistency, as its not present in other echart charts.
4. Series Limit -> has been changed to Row Limit as per review from @zhaoyongjie above to be inline with newer updates.
5. Max bubble size -> its there in customize section.
![Screenshot from 2022-12-14 07-44-19](https://user-images.githubusercontent.com/12967587/207488647-ad55b175-9c59-449b-8b75-e8baf978727b.png)


I tested above on ephemeral deployment.

rusackas commented:
@mayurnewase this appears to need a rebase. It might also need some commits added that change the old bubble chart name to include "legacy." @villebro has done some of this type of renaming and can probably warn us of any intricacies here.

EugeneTorap commented:
Hi @mayurnewase! Do you have some time to finish the PR?

mayurnewase commented:
@EugeneTorap unfortunately no, feel free to take over or close this PR.

john-bodley commented:
cc @michael-s-molina for context who has been working on other EChart migrations.

michael-s-molina commented:
Hi @mayurnewase. Thank you for this contribution. I would like to continue the work and see if we can ship this feature. To do that I can push to your PR if you give me the GitHub permission or I can close this PR and open my own and reference you as the original author. Which method do you prefer?

mayurnewase commented:
Hey thanks for taking this up, I am ok with both methods.
Added you as a collaborator in my fork of this repo, and I see ` Maintainers are allowed to edit this pull request.` in the right panel.

michael-s-molina commented:
> Added you as a collaborator in my fork of this repo, and I see  Maintainers are allowed to edit this pull request. in the right panel.

Thanks @mayurnewase. I'll rebase the PR and modify its scope a little bit to avoid introducing a breaking change. I'll move the migration script to its own PR, similar to https://github.com/apache/superset/pull/23973 and also keep the legacy plugin. The new scope will be to just add the ECharts Bubble chart which will allow us to merge this PR without the need to wait for a major version release.

michael-s-molina committed:
Removes migration script

michael-s-molina committed:
Preserves legacy plugin

michael-s-molina committed:
Adds thumbnails

michael-s-molina committed:
Fixes lint errors

michael-s-molina committed:
Changes default Y-axis margin

michael-s-molina commented:
I deprecated the legacy chart and added thumbnails for the new version.

<img width="1063" alt="Screenshot 2023-09-20 at 13 41 49" src="https://github.com/apache/superset/assets/70410625/07a82cce-c358-483f-abae-5e76a6decc50">

michael-s-molina commented:
/testenv up

villebro reviewed (approved):
I think this looks great. The only nit that comes to mind is different handling of the title and missing colors in the tooltip:

legacy:
![image](https://github.com/apache/superset/assets/33317356/0075036d-f6a5-4a65-80b6-a8a0d0fcb311)

new:
![image](https://github.com/apache/superset/assets/33317356/51b7c22d-a8d2-4d79-a7d7-eb7edeae389f)

But as discussed a few months ago, I think we have some tooltip harmonization and re-styling work to do anyway, so maybe this can be deferred to that effort.

michael-s-molina commented:
Thanks for the review @villebro!

> But as discussed a few months ago, I think we have some tooltip harmonization and re-styling work to do anyway, so maybe this can be deferred to that effort.

Yes, let's tackle this in a follow-up as part of the harmonization and re-styling work.
